tests/conftest.py

```diff
@@ -0,0 +1,54 @@
+"""Ruff is forcing me to write a docstring for conftest.py."""
+
+import shutil
+from typing import TYPE_CHECKING, Generator
+
+import pytest
+from _pytest.monkeypatch import MonkeyPatch
+
+if TYPE_CHECKING:
+    from _pytest.config.argparsing import Parser
+    from _pytest.tmpdir import TempPathFactory
+
+
+def pytest_addoption(parser: "Parser") -> None:
+    """Add options to pytest."""
+    parser.addoption(
+        "--reuse-envs",
+        action="store_true",
+        help="After tests run, don't remove hatch environments created for "
+        "generated projects (not the test environments for the "
+        "pyos-package-template project itself).\n"
+        "otherwise, use a temporary directoy and remove it afterwards.",
+    )
+
+
+@pytest.fixture(scope="session")
+def monkeypatch_session() -> Generator[MonkeyPatch, None, None]:
+    """Monkeypatch you can use with a session scoped fixture."""
+    mpatch = MonkeyPatch()
+    yield mpatch
+    mpatch.undo()
+
+
+@pytest.fixture(scope="session", autouse=True)
+def cleanup_hatch_envs(
+    pytestconfig: pytest.Config,
+    tmp_path_factory: "TempPathFactory",
+    monkeypatch_session: MonkeyPatch,
+) -> None:
+    """
+    Use a temporary directory for hatch envs & cleanup after session.
+
+    Unless --reuse-envs flag present.
+    """
+    if pytestconfig.getoption("--reuse-envs"):
+        yield
+        return
+
+    hatch_dir = tmp_path_factory.mktemp("hatch")
+    monkeypatch_session.setenv("HATCH_DATA_DIR", str(hatch_dir))
+
+    yield
+
+    shutil.rmtree(hatch_dir, ignore_errors=True)
```

Comment:
it can be if you'd like! 

`ignore_errors` makes the `onexc` call here be just a `pass`, so it handles these errors: https://github.com/python/cpython/blob/43447cb63421fb4db5411c1e74bdd8a4cfcf9263/Lib/shutil.py#L670-740 

but if you'd like to add a catchall here i don't see the harm in that

Reply:
I meant
```
try:
  yield
finally:
  shutil
```

Reply:
aha, i have never seen that pattern before, i think pytest catches the error and then goes back and calls `next()` for all the fixtures, so any error that would happen in the test wouldn't stop the cleanup, but i think even if it did this would still be fine so np lemme try that out

Reply:
there it is :) https://github.com/pyOpenSci/pyos-package-template/pull/65/commits/86b8c6643b8538008b1ad17009b1ab59d6af3b3d

Reply:
I have always used that pattern when testing with databases and rolling back transactions. I would really be interested to know if it's unnecessary.

Reply:
i think if there was some unhandled exception within pytest itself that caused it to lose the reference to the called fixture object, but didn't completely crash the interpreter, that it would save you there? idk give it a shot, i think there's no reason not to do it, since the next thing after the yield is the finally anyway